### ISSUE TITLE
Error in reading files from SPIFFS

### ISSUE THREAD

**Issue by shreyaspaqs** [espressif/arduino-esp32#2017]
### Hardware:
Board:							ESP32 Dev Module
Core Installation/update date:			08/june/2016
IDE name:							Arduino IDE
Flash Frequency:					40Mhz
PSRAM enabled:                                         no
Upload Speed:						115200
Computer OS:                                            Windows 10

### Description:
Whenever network is down, storing the data files(individual file for a packet) to flash memory(SPIFFS), when the network is up again, have to read all the files that are stored in flash memory. But, while reading the files, getting only the latest file repeated for number of times that are stored in memory.
Also, not able to pass a variable for filename(which has to be updated for every iteration)

```cpp

//Change the code below by your sketch
#include <Arduino.h>

void setup() {
  Serial.begin(115200);
  delay(1000);
  Serial.println("Starting SPI FLASH");
  connect_wifi();
  connect_mqtt();
  delay(1000);
  SPIFFS.begin();
  
   if(!SPIFFS.begin(FORMAT_SPIFFS_IF_FAILED)){
        Serial.println("SPIFFS Mount Failed");
        return;
    }else{
        Serial.println("SPIFFS Mount Success");
    }
}

void loop() {
  float temp1,rh,o3;
//  int RI=0,WI=0;
  int co2,tvoc,pm2_5,pm10;
  String WIstring = String(WI);
  String RIstring = String(RI);
  String esp32_reading="{ \"temp\" : " + WIstring + " ,  \"RH\" : " + RIstring + " }";
  RAW_DATA = esp32_reading.c_str();
  Serial.println(esp32_reading);
  delay(1000);
  client.publish("r/testdata", RAW_DATA );
  Serial.println("published");
  Serial.println("\n");
  if((WiFi.status() != WL_CONNECTED) || (client.connected() == MQTT_CONNECTED)){
    connect_wifi();
   // delay(10000);
   if(WiFi.status() == WL_CONNECTED) {
   connect_mqtt();
   if(client.connected() != MQTT_CONNECTED){
    while(WI != RI){
      char filename[20];
  String convertr;
  listDir(SPIFFS, "/", 0);
  //RI = 0;
  convertr = String(RI);
  Serial.print(RI);
  convertr.toCharArray(filename, 10);
  strcat(filename,".TXT");
  Serial.print(filename);
      Serial.println("Reading\n");
      
    //read_data();
    readFile(SPIFFS, "/filename");
    delay(1000);
    RI = RI+1;
    }
    Serial.println("skipped");
   }
   }else{
  Serial.println("Start storing data to a file \n");
  int w_stat,r_stat;
  char filename[20];
  String convert;
  convert = String(WI);
  convert.toCharArray(filename, 10);
  strcat(filename,".TXT");
   w_stat = writeFile(SPIFFS, "/filename", RAW_DATA);
   Serial.print(filename);
   if(w_stat == 1){
     WI++;
     Serial.print(WI);
     Serial.println("\n");
     Serial.print(RI);
   }else{
    writeFile(SPIFFS, "/filename", RAW_DATA);
   }
    Serial.println( "write complete" );
   }
  } 
}

void readFile(fs::FS &fs, char * path){
  
  String rconvert1; 
    Serial.printf("Reading file: %s\r\n", path);
    Serial.print(filename);
    File file2 = fs.open(path,"r");
    if(!file2 || file2.isDirectory()){
        Serial.println("- failed to open file for reading");
        return;
      }
    Serial.println("- read from file:");
    while(file2.available()){
   //   for(int j=0;j<=WI;j++){
      rconvert1 = file2.readString();
      STO_DATA[j] = rconvert1.c_str();
      Serial.println("reading as ");
        Serial.print(file2.read());
        Serial.println("\n");
        client.publish("paqs/myair/testdata", STO_DATA[j]);
        Serial.write(STO_DATA[j]);
        Serial.println("read as ");
        Serial.println("\n");
        j++;
        file2.close();
    //    return;
    }
    }
 //    return 1;
//}

int writeFile(fs::FS &fs, const char * path, const char * message){
    Serial.printf("Writing file: %s\r\n", path);
    File file = fs.open(path, FILE_WRITE);
    if(!file){
        Serial.println("- failed to open file for writing");
        return 0;
    }else{
     // for(int i=0;i<=5;i++){
        STO_DATA[i] = message;
        Serial.print(message);
        Serial.println("written like \n");
        Serial.print(STO_DATA[i]);
        i++;
     // }
    }
    if(file.print(message)){
        Serial.println("- file written");
        file.close();
        return 1;
    } else {
        Serial.println("- fwrite failed");
        return 0;
    }
}

```
### Debug Messages:
```
Enable Core debug level: none
output:

{ "temp" : 0 ,  "RH" : 0 }
published


{ "temp" : 0 ,  "RH" : 0 }
published

Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
{ "temp" : 0 ,  "RH" : 0}written like 

{ "temp" : 0 ,  "RH" : 0 }- file written
01

0write complete
{ "temp" : 1 ,  "RH" : 0 }
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
{ "temp" : 1 ,  "RH" : 0 }written like 

{ "temp" : 1 ,  "RH" : 0 }- file written
12

0write complete
{ "temp" : 2 ,  "RH" : 0}
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
{ "temp" : 2 ,  "RH" : 0 }written like 

{ "temp" : 2 ,  "RH" : 0}- file written
23

0write complete
{ "temp" : 3 ,  "RH" : 0 }
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
{ "temp" : 3 ,  "RH" : 0 }written like 

{ "temp" : 3 ,  "RH" : 0 }- file written
34

0write complete
{ "temp" : 4 ,  "RH" : 0}
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
Writing file: /wfilename
{ "temp" : 4 ,  "RH" : 0 }written like 

{ "temp" : 4 ,  "RH" : 0}- file written
78

0write complete
{ "temp" : 5 ,  "RH" : 0 }
published
45

0write complete
{ "temp" : 5 ,  "RH" : 0,  "tVOC" : 3,  "co2" : 4, "o3" : 5,  "PM2_5" : 6,  "PM10" : 7,  "devid" : 8 }
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
Writing file: /wfilename
{ "temp" : 5 ,  "RH" : 0 }written like 

{ "temp" : 5 ,  "RH" : 0}- file written
78

0write complete
{ "temp" : 8 ,  "RH" : 0 }
published

0write complete
{ "temp" : 6 ,  "RH" : 0,  "tVOC" : 3,  "co2" : 4, "o3" : 5,  "PM2_5" : 6,  "PM10" : 7,  "devid" : 8 }
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
Writing file: /wfilename
{ "temp" : 6 ,  "RH" : 0 }written like 

{ "temp" : 6 ,  "RH" : 0}- file written
78

0write complete
{ "temp" : 8 ,  "RH" : 0 }
published

0write complete
{ "temp" : 7 ,  "RH" : 0,  "tVOC" : 3,  "co2" : 4, "o3" : 5,  "PM2_5" : 6,  "PM10" : 7,  "devid" : 8 }
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
not Connected to the WiFi network
Connected to the WiFi network
Start storing data to a file 

Writing file: /wfilename
{ "temp" : 7 ,  "RH" : 0 }written like 

{ "temp" : 7 ,  "RH" : 0}- file written
78

0write complete
{ "temp" : 8 ,  "RH" : 0 }
published


Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connecting to WiFi..
Connected to the WiFi network
Connecting to MQTT...
connected
Listing directory: /
  FILE: /wfilename	SIZE: 102
0.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
1.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
2.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
3.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
4.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
5.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
6.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
Listing directory: /
  FILE: /wfilename	SIZE: 102
7.TXT.TXT.TXT.TXT.TXT.TXT.TXT.TXTReading

Reading file: /rfilename
/rfilename- failed to open file for reading
skipped
{ "temp" : 8 ,  "RH" : 8}
published


{ "temp" : 8 ,  "RH" : 8}
published

```

**Comment by lbernstone** [comment]
TL;DR.  This is not a code review site.  If you have an error with the codebase, post code that shows the error and only the error. FILE_WRITE starts at pos 0.  FILE_APPEND starts at the end of the file.

**Comment by shreyaspaqs** [comment]
Thanx lbernstone, 

How to pass a variable as a filename, so that every message should to be stored in a different file?

writeFile(SPIFFS, "/filename", RAW_DATA);  // "/filename " becomes the name of file

char filename[20];
  String convert;
  convert = String(WI);  //WI keeps on updated for each message
  convert.toCharArray(filename, 10);
  strcat(filename,".TXT");
example: Serial.print(filename);  //1.TXT
writeFile(SPIFFS, filename, RAW_DATA);   //throws error

can u help me out in solving this issue?
